vulnerabilities/package_managers.py

```diff
@@ -489,3 +509,75 @@ async def fetch(self, pkg, session):
             pass
 
         self.cache[pkg] = versions
+
+
+class GoproxyVersionAPI(VersionAPI):
+
+    package_type = "golang"
+    pkg_mappings = {}
+
+    @staticmethod
+    def trim_package_path(pkg: str) -> Optional[str]:
+        """github advisories for golang is using package names(e.g. https://github.com/advisories/GHSA-jp4j-47f9-2vc3), yet goproxy works with module names(see https://golang.org/ref/mod#goproxy-protocol).
+        this method removes the last part of a package path, and returns the remaining as the module name.
+        """
+        # some advisories contains this prefix in package name, e.g. https://github.com/advisories/GHSA-7h6j-2268-fhcm
+        pkg = pkg.removeprefix("https://pkg.go.dev/")
+        parts = pkg.split("/")
+        if len(parts) >= 2:
+            return "/".join(parts[:-1])
+        else:
+            return None
+
+    @staticmethod
+    def escape_path(path: str) -> str:
+        """escpe uppercase in module/version name"""
+        escaped_path = ""
+        for c in path:
+            if c >= "A" and c <= "Z":
+                escaped_path += "!" + chr(ord(c) + ord("a") - ord("A"))
+            else:
+                escaped_path += c
+        return escaped_path
+
+    async def fetch(self, pkg: str, session: ClientSession):
+        # escpe uppercase in module path
+        escaped_pkg = GoproxyVersionAPI.escape_path(pkg)
+        trimmed_pkg = pkg
+        resp_text = None
+        err_pkgs = []
+        while escaped_pkg is not None:
```

Comment:
I don't understand this, why is there a loop ? Maybe simplify or add a comment  on this ?

Reply:
My guess is that this code tries to find some package on the proxy.golang thingy if it doesn't find it, some part of package is chopped. This is done till a match is found or everything is chopped.

Reply:
yes you're right. github advisory contains package path,  proxy.golang only works with module path and will report error when missing with http status 410.

Reply:
Alright. Please add a comment for this. Otherwise we'll forget what's this doing ;)

Reply:
Maybe just provide a link to this https://go.dev/ref/mod#resolve-pkg-mod